### ISSUE TITLE
[LLVM18 Packaging] Incorrect symlinks to shared libraries

### ISSUE THREAD

**Issue by kledom** [llvm/llvm-project#83986]
**Description:**
I'm compiling include-what-you-use against LLVM 18 inside a docker container. The llvm-toolchain is installed using apt. Since Monday I am seeing errors when running iwyu:

```
> include-what-you-use --version
include-what-you-use: error while loading shared libraries: libclang-cpp.so.18.1: cannot open shared object file: No such file or directory
```

The problem is probably that the symlinks within `/usr/lib/x86_64-linux-gnu/` have to be adapted for version 18.1:
```
dpkg --contents ./libclang-cpp18_1%3a18.1.0~++20240302042158+461274b81d86-1~exp1~20240302162229.64_amd64.deb 
drwxr-xr-x root/root         0 2024-03-02 16:22 ./
drwxr-xr-x root/root         0 2024-03-02 16:22 ./usr/
drwxr-xr-x root/root         0 2024-03-02 16:22 ./usr/lib/
drwxr-xr-x root/root         0 2024-03-02 16:22 ./usr/lib/llvm-18/
drwxr-xr-x root/root         0 2024-03-02 16:22 ./usr/lib/llvm-18/lib/
-rw-r--r-- root/root  65946936 2024-03-02 16:22 ./usr/lib/llvm-18/lib/libclang-cpp.so.18.1
drwxr-xr-x root/root         0 2024-03-02 16:22 ./usr/lib/x86_64-linux-gnu/
drwxr-xr-x root/root         0 2024-03-02 16:22 ./usr/share/
drwxr-xr-x root/root         0 2024-03-02 16:22 ./usr/share/doc/
drwxr-xr-x root/root         0 2024-03-02 16:22 ./usr/share/doc/libclang-cpp18/
-rw-r--r-- root/root     64007 2024-03-02 16:22 ./usr/share/doc/libclang-cpp18/changelog.Debian.gz
-rw-r--r-- root/root      6897 2024-03-02 16:21 ./usr/share/doc/libclang-cpp18/copyright
lrwxrwxrwx root/root         0 2024-03-02 16:22 ./libclang-cpp.so.18 -> ../llvm-18/lib/libclang-cpp.so.18
```

**Affected packages:**

```
apt-cache show llvm-18-dev | grep "Version"
Version: 1:18.1.0~++20240302042158+461274b81d86-1~exp1~20240302162229.64
```
```
apt-cache show libclang-cpp18 | grep "Version"
Version: 1:18.1.0~++20240302042158+461274b81d86-1~exp1~20240302162229.64
```

**Comment by TheRealCuran** [comment]
Sort of a duplicate of #82779 (though that ticket was submitted for the LLVM19/Snapshot packages).

**Comment by sylvestre** [comment]
yeah, i fixed it already in debian upstream. let me retrigger builds for apt.llvm.org

**Comment by kledom** [comment]
Unfortunately, there are still problems with the new build.

```
apt-cache show llvm-18-dev | grep "Version"
Version: 1:18.1.0~++20240307093122+461274b81d86-1~exp1~20240307213239.65
```

The path looks somewhat better now, but ../llvm-18/lib/libclang-cpp.so.18 still leads nowhere:

```
dpkg -c ./libclang-cpp18_18.1.0~++20240307093122+461274b81d86-1~exp1~20240307213239.65_amd64.deb 
drwxr-xr-x root/root         0 2024-03-07 21:32 ./
drwxr-xr-x root/root         0 2024-03-07 21:32 ./usr/
drwxr-xr-x root/root         0 2024-03-07 21:32 ./usr/lib/
drwxr-xr-x root/root         0 2024-03-07 21:32 ./usr/lib/llvm-18/
drwxr-xr-x root/root         0 2024-03-07 21:32 ./usr/lib/llvm-18/lib/
-rw-r--r-- root/root  65946936 2024-03-07 21:32 ./usr/lib/llvm-18/lib/libclang-cpp.so.18.1
[...]
lrwxrwxrwx root/root         0 2024-03-07 21:32 ./usr/lib/x86_64-linux-gnu/libclang-cpp.so.18 -> ../llvm-18/lib/libclang-cpp.so.18
```

**Comment by sylvestre** [comment]
ok, i will fix that in the next few hours.
Sorry, the recent changes of soname naming can be long to handle in debian packages

**Comment by kledom** [comment]
Thank you for your quick response.
Fixed with `18.1.1~++20240308111306+dba2a75e9c7e-1~exp1~20240308231321.67`:

```
dpkg-deb -c ./libclang-cpp18_18.1.1~++20240308111306+dba2a75e9c7e-1~exp1~20240308231321.67_amd64.deb 
drwxr-xr-x root/root         0 2024-03-08 23:13 ./
drwxr-xr-x root/root         0 2024-03-08 23:13 ./usr/
drwxr-xr-x root/root         0 2024-03-08 23:13 ./usr/lib/
drwxr-xr-x root/root         0 2024-03-08 23:13 ./usr/lib/llvm-18/
drwxr-xr-x root/root         0 2024-03-08 23:13 ./usr/lib/llvm-18/lib/
-rw-r--r-- root/root  65946936 2024-03-08 23:13 ./usr/lib/llvm-18/lib/libclang-cpp.so.18.1
drwxr-xr-x root/root         0 2024-03-08 23:13 ./usr/lib/x86_64-linux-gnu/
drwxr-xr-x root/root         0 2024-03-08 23:13 ./usr/share/
drwxr-xr-x root/root         0 2024-03-08 23:13 ./usr/share/doc/
drwxr-xr-x root/root         0 2024-03-08 23:13 ./usr/share/doc/libclang-cpp18/
-rw-r--r-- root/root     64289 2024-03-08 23:13 ./usr/share/doc/libclang-cpp18/changelog.Debian.gz
-rw-r--r-- root/root      6897 2024-03-08 23:12 ./usr/share/doc/libclang-cpp18/copyright
lrwxrwxrwx root/root         0 2024-03-08 23:13 ./usr/lib/x86_64-linux-gnu/libclang-cpp.so.18 -> ../llvm-18/lib/libclang-cpp.so.18.1
lrwxrwxrwx root/root         0 2024-03-08 23:13 ./usr/lib/x86_64-linux-gnu/libclang-cpp.so.18.1 -> ../llvm-18/lib/libclang-cpp.so.18.1
```
